include/lib/map.h

```diff
@@ -4,24 +4,39 @@
 
 #include "stdlib/stdtypes.h"
 
-/** Data structure for a generic map */
+/** Structures for a generic map. */
+struct entry {
+    /** The key associated with the entry. */
+    void *key;
+    /** The value associated with the entry. */
+    void *value;
+};
+
 struct map {
-    //TODO
+    /** Table which holds all entries. */
+    struct entry *table;
+    /** Amount of space used. */
+    size_t used;
+    /** Maximum space to allocate for entries */
+    size_t size;
 };
 
-/** Initializes a new hashtable */
-void map_init(struct map *m);
+/** Initializes a new hashmap. Returns `-1` if not allocated. */
+int map_init(struct map *m, size_t size);
 
-/** Frees the hashtable from memory */
+/** Frees the hashtable from memory. */
 void map_free(struct map *m);
 
-/** Adds the key/value pair to the hashtable */
-void map_put(struct map *m, void *key, void *value);
+/** Adds the key/value pair to the hashtable. */
+int map_put(struct map *m, void *key, void *value, size_t size);
+
+/** Gets the value associated with the specified key. */
+void *map_get(struct map *m, void *key, size_t size);
 
-/** Gets the value associated with the specified key */
-void *map_get(struct map *m, void *key);
+/** Checks if key exists in hashmap. */
+bool map_contains(struct map *m, void *key, size_t size);
 
-/** Returns true if the hashtable contains the key, false otherwise */
-bool map_contains(struct map *m, void *key);
```

Comment:
You should still create `map_contains` even if you think it's redundant. I know that it's redundant but using boolean getters can make code easier to read down the line. You should reimplement this. You're gonna want to run any changes to the standard library API by me first.